File: VB/ScheduledTasksAPIClientDemo/ConnectionProviderExtensions.vb
Imports DevExpress.ReportServer.ServiceModel.Client
Imports DevExpress.ReportServer.ServiceModel.ConnectionProviders
Imports System
Imports System.Threading.Tasks

Namespace ScheduledTasksAPIClientDemo
    Public Module ConnectionProviderExtensions
        <System.Runtime.CompilerServices.Extension> _
        Public Function DoWithScheduledJobAsync(Of T)(ByVal serverConnection As ConnectionProvider, ByVal func As Func(Of IReportServerClient, Task(Of T))) As Task(Of T)
            Return serverConnection.ConnectAsync().ContinueWith(Function(taskFunc)
                Dim client As IReportServerClient = taskFunc.Result
                Return func(client)
            End Function).Unwrap()
        End Function

        <System.Runtime.CompilerServices.Extension> _
        Public Function DoWithScheduledJobAsync(ByVal serverConnection As ConnectionProvider, ByVal func As Func(Of IReportServerClient, Task)) As Task
            Return serverConnection.ConnectAsync().ContinueWith(Function(task) func(task.Result)).Unwrap()
        End Function
    End Module
End Namespace

File: CS/ScheduledTasksAPIClientDemo/ConnectionProviderExtensions.cs
using DevExpress.ReportServer.ServiceModel.Client;
using DevExpress.ReportServer.ServiceModel.ConnectionProviders;
using System;
using System.Threading.Tasks;

namespace ScheduledTasksAPIClientDemo {
    public static class ConnectionProviderExtensions {
        public static Task<T> DoWithScheduledJobAsync<T>(this ConnectionProvider serverConnection, Func<IReportServerClient, Task<T>> func) {
            return serverConnection
                .ConnectAsync()
                .ContinueWith(taskFunc => {
                    IReportServerClient client = taskFunc.Result;
                    return func(client);
                })
                .Unwrap();
        }

        public static Task DoWithScheduledJobAsync(this ConnectionProvider serverConnection, Func<IReportServerClient, Task> func) {
            return serverConnection
                .ConnectAsync()
                .ContinueWith(task => func(task.Result))
                .Unwrap();
        }
    }
}
